extensions/sql-database-projects/src/test/mainController.test.ts

```diff
@@ -25,53 +24,50 @@ describe('MainController: main controller operations', function (): void {
 		await baselines.loadBaselines();
 	});
 
-	beforeEach(function (): void {
-		testContext.apiWrapper.reset();
+	afterEach(function (): void {
+		sinon.restore();
 	});
 
 	it('Should create new project through MainController', async function (): Promise<void> {
 		const projFileDir = path.join(os.tmpdir(), `TestProject_${new Date().getTime()}`);
 
-		testContext.apiWrapper.setup(x => x.showInputBox(TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve('MyProjectName'));
-		testContext.apiWrapper.setup(x => x.showOpenDialog(TypeMoq.It.isAny())).returns(() => Promise.resolve([vscode.Uri.file(projFileDir)]));
-		testContext.apiWrapper.setup(x => x.workspaceFolders()).returns(() => undefined);
-		testContext.apiWrapper.setup(x => x.showErrorMessage(TypeMoq.It.isAny())).returns((s) => {
-			console.log(s);
-			return Promise.resolve(s);
-		});
+		sinon.stub(vscode.window, 'showInputBox').returns(Promise.resolve('MyProjectName'));
+		sinon.stub(vscode.window, 'showOpenDialog').returns(Promise.resolve([vscode.Uri.file(projFileDir)]));
+		sinon.replaceGetter(vscode.workspace, 'workspaceFolders', () => undefined);
 
-		const controller = new MainController(testContext.context, testContext.apiWrapper.object);
+		const controller = new MainController(testContext.context);
 		const proj = await controller.createNewProject();
 
 		should(proj).not.equal(undefined);
 	});
 
 	it('Should show error when no project name', async function (): Promise<void> {
 		for (const name of ['', '    ', undefined]) {
-			testContext.apiWrapper.reset();
-			testContext.apiWrapper.setup(x => x.showInputBox(TypeMoq.It.isAny(), TypeMoq.It.isAny())).returns(() => Promise.resolve(name));
-			testContext.apiWrapper.setup(x => x.showErrorMessage(TypeMoq.It.isAny())).returns((s) => { throw new Error(s); });
-
-			const controller = new MainController(testContext.context, testContext.apiWrapper.object);
-			await shouldThrowSpecificError(async () => await controller.createNewProject(), constants.projectNameRequired, `case: '${name}'`);
+			const stub = sinon.stub(vscode.window, 'showInputBox').returns(Promise.resolve(name));
+			const spy = sinon.spy(vscode.window, 'showErrorMessage');
+			const controller = new MainController(testContext.context);
+			await controller.createNewProject();
+			should(spy.calledOnce).be.true('showErrorMessage should have been called exactly once');
+			stub.restore();
+			spy.restore();
```

Comment:
when does spy.restore() need to be called? I noticed it isn't called at the end in other places that also declare it as a variable

Reply:
Calling stub/spy on a function that was already wrapped will throw an error. So you need to restore it before trying to wrap it again. In this case the loop causes the function to be wrapped multiple times.

I could have also declared the spy outside the loop and then called .reset at the end of every loop. I just wasn't paying close enough attention to minor optimizations like that when I was writing this (and personally don't have a huge preference either way, both are fine). But it needs to be reset in order to reset the call count.